receiver/splunkhecreceiver/receiver.go

```diff
@@ -0,0 +1,319 @@
+			//		set some hard-coded values.
+			ReadHeaderTimeout: defaultServerTimeout,
+			WriteTimeout:      defaultServerTimeout,
+		},
+	}
+
+	return r, nil
+}
+
+// StartMetricsReception tells the receiver to start its processing.
+// By convention the consumer of the received data is set when the receiver
+// instance is created.
+func (r *splunkReceiver) Start(_ context.Context, host component.Host) error {
+	r.Lock()
+	defer r.Unlock()
+
+	err := componenterror.ErrAlreadyStarted
+	r.startOnce.Do(func() {
+		err = nil
+
+		var ln net.Listener
+		// set up the listener
+		ln, err = r.config.HTTPServerSettings.ToListener()
+		if err != nil {
+			err = fmt.Errorf("failed to bind to address %s: %w", r.config.Endpoint, err)
+			return
+		}
+
+		mx := mux.NewRouter()
+		mx.HandleFunc(hecPath, r.handleReq)
+
+		r.server = r.config.HTTPServerSettings.ToServer(mx)
+
+		// TODO: Evaluate what properties should be configurable, for now
+		//		set some hard-coded values.
+		r.server.ReadHeaderTimeout = defaultServerTimeout
+		r.server.WriteTimeout = defaultServerTimeout
+
+		go func() {
+			if errHTTP := r.server.Serve(ln); errHTTP != nil {
+				host.ReportFatalError(errHTTP)
+			}
+		}()
+	})
+
+	return err
+}
+
+// StopMetricsReception tells the receiver that should stop reception,
+// giving it a chance to perform any necessary clean-up.
+func (r *splunkReceiver) Shutdown(context.Context) error {
+	r.Lock()
+	defer r.Unlock()
+
+	err := componenterror.ErrAlreadyStopped
+	r.stopOnce.Do(func() {
+		err = r.server.Close()
+	})
+	return err
+}
+
+func (r *splunkReceiver) handleReq(resp http.ResponseWriter, req *http.Request) {
+	transport := "http"
+	if r.config.TLSSetting != nil {
+		transport = "https"
+	}
+	ctx := obsreport.ReceiverContext(req.Context(), r.config.Name(), transport, r.config.Name())
+	ctx = obsreport.StartMetricsReceiveOp(ctx, r.config.Name(), transport)
+
+	if req.Method != http.MethodPost {
+		r.failRequest(ctx, resp, http.StatusBadRequest, invalidMethodRespBody, nil)
+		return
+	}
+
+	if req.Header.Get(httpContentTypeHeader) != jsonContentType {
+		r.failRequest(ctx, resp, http.StatusUnsupportedMediaType, invalidContentRespBody, nil)
+		return
+	}
+
+	encoding := req.Header.Get(httpContentEncodingHeader)
+	if encoding != "" && encoding != gzipEncoding {
+		r.failRequest(ctx, resp, http.StatusUnsupportedMediaType, invalidEncodingRespBody, nil)
+		return
+	}
+
+	bodyReader := req.Body
+	if encoding == gzipEncoding {
+		var err error
+		bodyReader, err = gzip.NewReader(bodyReader)
+		if err != nil {
+			r.failRequest(ctx, resp, http.StatusBadRequest, errGzipReaderRespBody, err)
+			return
+		}
+	}
+
+	body, err := ioutil.ReadAll(bodyReader)
+	if err != nil {
+		r.failRequest(ctx, resp, http.StatusBadRequest, errReadBodyRespBody, err)
+		return
+	}
+	if len(body) == 0 {
+		resp.Write(okRespBody)
+		return
+	}
+
+	messagesReceived := 0
+	dec := json.NewDecoder(bytes.NewReader(body))
```

Comment:
Why ReadAll instead of processing by streaming here?

Reply:
Good point, it's a remnant of when I was splitting the string. Removed.